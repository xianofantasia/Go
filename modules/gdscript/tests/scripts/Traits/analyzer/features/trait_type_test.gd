trait Shearable:
	func shear() -> void:
		print("Shearable")

trait Milkable:
	func milk() -> void:
		print("Milkable")

class FluffyCow:
	uses Shearable, Milkable

class FluffyBull:
	uses Shearable

class Bird:
	pass

func test_func(node: Object):
	if node is Shearable:
		print("Shearable")
	if node is Milkable:
		print("Milkable")

func test():
	print(FluffyCow is Shearable)
	print(FluffyCow is Milkable)

	var bull = FluffyBull.new()
	print(bull is Shearable)
	print(bull is Milkable)
	test_func(bull)

	var my_animals : Array = []
	my_animals.append(FluffyCow.new())
	my_animals.append(FluffyBull.new())
	my_animals.append(Bird.new())
	var count = 1
	for animal in my_animals:
		print("Animal ", count)
		if animal is Shearable:
			animal.shear()
		if animal is Milkable:
			animal.milk()
		count += 1

	print("ok")
