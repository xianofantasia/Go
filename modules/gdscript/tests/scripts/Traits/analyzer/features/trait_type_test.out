GDTEST_OK
~~ WARNING at line 31: (UNSAFE_CALL_ARGUMENT) The argument 1 of the function "test_func()" requires the subtype "Variant" but the supertype "Variant" was provided.
~~ WARNING at line 41: (UNSAFE_METHOD_ACCESS) The method "shear()" is not present on the inferred type "Variant" (but may be present on a subtype).
~~ WARNING at line 43: (UNSAFE_METHOD_ACCESS) The method "milk()" is not present on the inferred type "Variant" (but may be present on a subtype).
true
true
true
false
Shearable
Animal 1
Shearable
Milkable
Animal 2
Shearable
Animal 3
ok
